File: apps/opensible-server/services/remote_state_lock.py
"""Remote state lock for preventing concurrent mutations across workers (UC331).

Each stack using a remote backend (S3/OBS) gets a lock keyed by (stack, backend_type, backend_key).
The lock is stored in PostgreSQL and uses advisory locks for atomicity.
"""
from __future__ import annotations

import time
import uuid
from typing import Any, Optional

from storage import pg

LOCK_PREFIX = "radas.remote_state_lock:"


def _lock_key(stack: str, backend_type: str, backend_key: str) -> str:
    return f"{LOCK_PREFIX}{stack}:{backend_type}:{backend_key}"


def acquire(stack: str, backend_type: str, backend_key: str, *, actor: str, operation: str, run_id: Optional[str] = None) -> dict[str, Any]:
    """Acquire a remote-state advisory lock and record the lease."""
    lock_key = _lock_key(stack, backend_type, backend_key)
    now = time.time()
    lease_id = str(uuid.uuid4())

    with pg.transaction() as conn:
        # Advisory lock (transaction-level)
        conn.execute("SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))", (lock_key,))

        # Check existing active lease
        existing = conn.execute(
            "SELECT * FROM remote_state_locks WHERE stack = %s AND backend_type = %s AND backend_key = %s AND expires_at > %s FOR UPDATE",
            (stack, backend_type, backend_key, now),
        ).fetchone()
        if existing:
            return {"ok": False, "lock": dict(existing)}

        expires_at = now + 3600  # 1 hour
        conn.execute(
            "INSERT INTO remote_state_locks (id, stack, backend_type, backend_key, actor, operation, run_id, acquired_at, expires_at) "
            "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)",
            (lease_id, stack, backend_type, backend_key, actor, operation, run_id, now, expires_at),
        )
        return {"ok": True, "lock": {"id": lease_id, "stack": stack, "backend_type": backend_type,
                                     "backend_key": backend_key, "actor": actor,
                                     "operation": operation, "run_id": run_id,
                                     "acquired_at": now, "expires_at": expires_at}}


def release(stack: str, backend_type: str, backend_key: str, *, lock_id: Optional[str] = None, force: bool = False) -> dict[str, Any]:
    """Release the remote state lock."""
    with pg.transaction() as conn:
        conn.execute("SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))", (_lock_key(stack, backend_type, backend_key),))
        if lock_id:
            result = conn.execute(
                "DELETE FROM remote_state_locks WHERE stack = %s AND backend_type = %s AND backend_key = %s AND id = %s RETURNING *",
                (stack, backend_type, backend_key, lock_id),
            )
        else:
            result = conn.execute(
                "DELETE FROM remote_state_locks WHERE stack = %s AND backend_type = %s AND backend_key = %s RETURNING *",
                (stack, backend_type, backend_key),
            )
        if result.rowcount == 0 and not force:
            return {"ok": False, "error": "No active lock found"}
        return {"ok": True, "released": result.rowcount > 0, "previous": result.fetchone()}


def get_lock(stack: str, backend_type: str, backend_key: str) -> Optional[dict[str, Any]]:
    now = time.time()
    row = pg.query_one(
        "SELECT * FROM remote_state_locks WHERE stack = %s AND backend_type = %s AND backend_key = %s AND expires_at > %s",
        (stack, backend_type, backend_key, now),
    )
    return dict(row) if row else None


def cleanup_expired() -> int:
    now = time.time()
    with pg.transaction() as conn:
        result = conn.execute("DELETE FROM remote_state_locks WHERE expires_at <= %s", (now,))
        return result.rowcount

File: apps/server/lib/radas_ai/remote_state_lock.ex
defmodule RadasAI.RemoteStateLock do
  @moduledoc """
  Port of `services/remote_state_lock.py` (UC331) — advisory lock for
  stacks on remote backends, keyed (stack, backend_type, backend_key),
  stored in the `remote_state_locks` table with 1h expiry.
  """

  import RadasAI.DB

  alias Radas.Repo

  @lock_prefix "radas.remote_state_lock:"

  defp lock_key(stack, backend_type, backend_key),
    do: "#{@lock_prefix}#{stack}:#{backend_type}:#{backend_key}"

  @doc "Acquire the remote-state lease: %{\"ok\" => bool, \"lock\" => lease}."
  @spec acquire(String.t(), String.t(), String.t(), keyword()) :: map()
  def acquire(stack, backend_type, backend_key, opts \\ []) do
    actor = Keyword.get(opts, :actor, "")
    operation = Keyword.get(opts, :operation, "")
    run_id = Keyword.get(opts, :run_id)
    now = now()
    lease_id = Ecto.UUID.generate()

    {:ok, result} =
      Repo.transaction(fn ->
        Repo.query!("SELECT pg_advisory_xact_lock(hashtextextended($1, 0))", [
          lock_key(stack, backend_type, backend_key)
        ])

        existing =
          query_one!(
            """
            SELECT * FROM remote_state_locks
            WHERE stack = $1 AND backend_type = $2 AND backend_key = $3 AND expires_at > $4
            FOR UPDATE
            """,
            [stack, backend_type, backend_key, now]
          )

        if existing do
          %{"ok" => false, "lock" => existing}
        else
          expires_at = now + 3600

          execute!(
            """
            INSERT INTO remote_state_locks
              (id, stack, backend_type, backend_key, actor, operation, run_id, acquired_at, expires_at)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
            """,
            [lease_id, stack, backend_type, backend_key, actor, operation, run_id, now, expires_at]
          )

          %{
            "ok" => true,
            "lock" => %{
              "id" => lease_id,
              "stack" => stack,
              "backend_type" => backend_type,
              "backend_key" => backend_key,
              "actor" => actor,
              "operation" => operation,
              "run_id" => run_id,
              "acquired_at" => now,
              "expires_at" => expires_at
            }
          }
        end
      end)

    result
  end

  @doc "Release the remote-state lease: %{\"ok\", \"released\"?, \"error\"?}."
  @spec release(String.t(), String.t(), String.t(), keyword()) :: map()
  def release(stack, backend_type, backend_key, opts \\ []) do
    lock_id = Keyword.get(opts, :lock_id)
    force = Keyword.get(opts, :force, false)

    {:ok, result} =
      Repo.transaction(fn ->
        Repo.query!("SELECT pg_advisory_xact_lock(hashtextextended($1, 0))", [
          lock_key(stack, backend_type, backend_key)
        ])

        {sql, params} =
          if lock_id do
            {
              """
              DELETE FROM remote_state_locks
              WHERE stack = $1 AND backend_type = $2 AND backend_key = $3 AND id = $4 RETURNING *
              """,
              [stack, backend_type, backend_key, lock_id]
            }
          else
            {
              "DELETE FROM remote_state_locks WHERE stack = $1 AND backend_type = $2 AND backend_key = $3 RETURNING *",
              [stack, backend_type, backend_key]
            }
          end

        rows = query_all!(sql, params)

        if rows == [] and not force do
          %{"ok" => false, "error" => "No active lock found"}
        else
          %{"ok" => true, "released" => rows != [], "previous" => List.first(rows)}
        end
      end)

    result
  end

  @doc "Active lock for a backend identity, if any (Python get_lock)."
  @spec get_lock(String.t(), String.t(), String.t()) :: map() | nil
  def get_lock(stack, backend_type, backend_key) do
    query_one!(
      """
      SELECT * FROM remote_state_locks
      WHERE stack = $1 AND backend_type = $2 AND backend_key = $3 AND expires_at > $4
      """,
      [stack, backend_type, backend_key, now()]
    )
  end

  @doc "Remove expired leases; returns the number of rows removed."
  @spec cleanup_expired() :: integer()
  def cleanup_expired do
    execute!("DELETE FROM remote_state_locks WHERE expires_at <= $1", [now()]) || 0
  end
end
